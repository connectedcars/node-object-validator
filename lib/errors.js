function ValidationError(message, context = {}) {
  this.name = 'ValidationError'
  this.message = message
  if (context.key) {
    this.key = context.key
  }
  if (context.val) {
    this.val = context.val
  }
}
ValidationError.prototype = new Error()

module.exports = {
  ValidationError
}
