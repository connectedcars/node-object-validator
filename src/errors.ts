export interface ValidationErrorContext<T> {
  key: string
  value: T
}

export class ValidationError<T> extends Error {
  public key?: string
  public value?: T
  public constructor(message: string, context?: ValidationErrorContext<T>) {
    if (context) {
      message = `Field '${context.key}' ${message.charAt(0).toLowerCase() + message.slice(1)}`
    }
    super(message)
    this.key = context?.key
    this.value = context?.value
  }
}

export class RequiredError extends ValidationError<string> {}
export class NotStringError extends ValidationError<string> {}
export class WrongLengthError extends ValidationError<string> {}
export class NotDateError extends ValidationError<string> {}
export class NotRfc3339Error extends ValidationError<string> {}
export class NotExactStringError extends ValidationError<string> {}
export class NotFloatError extends ValidationError<string> {}
export class NotIntegerError extends ValidationError<string> {}
export class NotArrayError extends ValidationError<string> {}
export class NotObjectError extends ValidationError<string> {}
export class DoesNotMatchRegexError extends ValidationError<string> {}
export class OutOfRangeError extends ValidationError<string> {}
