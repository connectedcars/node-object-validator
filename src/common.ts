import { RequiredFail, ValidationFailure, ValidationsError } from './errors'

// https://stackoverflow.com/questions/51651499/typescript-what-is-a-naked-type-parameter
// https://2ality.com/2019/07/testing-static-types.html
// Wrapping the types in an tuple force a specific type instead of allow any in the union
export type AssertEqual<T, Expected> = [T, Expected] extends [Expected, T] ? true : false

export type ReturnEqual<T, C> = [T, C] extends [C, T] ? C : never

// https://github.com/microsoft/TypeScript/issues/41746
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type BaseObject = Record<string, any>

// TODO: Give better name
export function isValidType<T>(value: unknown, errors: ValidationFailure[]): value is T {
  return errors.length === 0
}

export type CodeGenResult = [{ [key: string]: unknown }, string[], string[]]

export type ValidatorBaseOptions = {
  optimize?: boolean
  required?: boolean
  nullable?: boolean
  earlyFail?: boolean
}

/**
 * @typedef ValidatorOptions
 * @property {boolean} [earlyFail=false] Stop validation on first failure (default: false)
 * @property {boolean} [optimize=true] Generate an optimized function for doing the validation (default: true)
 */
export interface ValidatorOptions {
  earlyFail?: boolean
  optimize?: boolean
}

export interface ValidateOptions {
  earlyFail?: boolean
  optimized?: boolean
}

export interface ValidatorExportOptions {
  language?: string
  types?: boolean
}

export function isValidator(value: unknown): value is ValidatorBase {
  if (value instanceof ValidatorBase) {
    return true
  }
  return false
}

export function generateOptionsString(options: ValidatorBaseOptions, defaults: Required<ValidatorBaseOptions>): string {
  const selectedOptions: string[] = []
  if (options.required !== undefined && options.required !== defaults.required) {
    selectedOptions.push(`required: ${options.required}`)
  }
  if (options.nullable !== undefined && options.nullable !== defaults.nullable) {
    selectedOptions.push(`nullable: ${options.nullable}`)
  }
  if (options.earlyFail !== undefined && options.earlyFail !== defaults.earlyFail) {
    selectedOptions.push(`earlyFail: ${options.earlyFail}`)
  }
  if (options.optimize !== undefined && options.optimize !== defaults.optimize) {
    selectedOptions.push(`optimize: ${options.optimize}`)
  }
  return selectedOptions.length > 0 ? `{ ${selectedOptions.join(', ')} }` : ''
}

export abstract class ValidatorBase<T = unknown> {
  public tsType!: T
  public required: boolean
  public nullable: boolean
  public earlyFail: boolean

  protected codeGenId = 1
  protected optimizedValidate: ((value: unknown, context?: string) => ValidationFailure[]) | null
  protected optionsString: string

  public constructor(options?: ValidatorBaseOptions) {
    const defaults = { required: true, nullable: false, earlyFail: false, optimize: true }
    const mergedOptions = { ...defaults, ...options }

    this.optionsString = options ? generateOptionsString(options, defaults) : ''

    this.required = mergedOptions.required
    this.earlyFail = mergedOptions.earlyFail
    this.nullable = mergedOptions.nullable
    this.optimizedValidate = null
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public AssertType<C extends T, X extends AssertEqual<T, C>>(): void {
    // Only used for type assertions
  }

  public isValid<C extends T>(obj: unknown): obj is ReturnEqual<T, C> {
    const errors = this.validate(obj)
    return errors.length === 0
  }

  public isType<C extends T>(obj: unknown, errors: ValidationFailure[]): obj is ReturnEqual<T, C> {
    return errors.length === 0
  }

  public cast<C extends T>(obj: unknown): ReturnEqual<T, C> {
    const errors = this.validate(obj)
    if (this.isType<C>(obj, errors)) {
      return obj
    } else {
      throw new ValidationsError('One of more validations failed', errors)
    }
  }

  public validate(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[] {
    if (options?.optimized !== false && this.optimizedValidate !== null) {
      return this.optimizedValidate(value, context)
    }
    if (this.nullable && value === null) {
      return []
    }
    if (value === undefined) {
      return this.required ? [new RequiredFail(`Is required`, value, context)] : []
    }
    return this.validateValue(value, context, { earlyFail: this.earlyFail, optimized: false, ...options })
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, { key: (context ? \`\${context}['${context}']\` : \`${context}\`) }` : ', context'
    const validatorName = `validator${id()}`
    const declarations = [`const ${validatorName} = ${validatorRef}`]
    // prettier-ignore
    const code: string[] = [
      ...this.nullCheckWrap([
        `  errors.push(...${validatorName}.validateValue(${valueRef}${contextStr}))`,
      ],valueRef, contextStr),
      ...(this.earlyFail || earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),
    ]
    return [
      {
        RequiredFail: RequiredFail
      },
      declarations,
      code
    ]
  }

  protected nullCheckWrap(code: string[], valueRef: string, contextStr: string, skip = false): string[] {
    if (skip) {
      return code
    }

    const result: string[] = []
    if (this.required) {
      if (this.nullable) {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(`  if (${valueRef} !== null) {`)
        result.push(...code)
        result.push(`  }`)
        result.push(`} else {`)
        result.push(`  errors.push(new RequiredFail(\`Is required\`, ${valueRef}${contextStr}))`)
        result.push(`}`)
      } else {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(...code)
        result.push(`} else {`)
        result.push(`  errors.push(new RequiredFail(\`Is required\`, ${valueRef}${contextStr}))`)
        result.push(`}`)
      }
    } else {
      if (this.nullable) {
        result.push(`if (${valueRef} != null) {`)
        result.push(...code)
        result.push(`}`)
      } else {
        result.push(`if (${valueRef} !== undefined) {`)
        result.push(...code)
        result.push(`}`)
      }
    }
    return result
  }

  protected optimize(schema?: unknown): void {
    const [imports, declarations, code] = this.codeGen(
      'value',
      'schema',
      () => {
        return this.codeGenId++
      },
      undefined,
      this.earlyFail
    )
    const functionBody = [
      ...Object.keys(imports).map(i => `const ${i} = imports['${i}']`),
      ...declarations,
      `return (value, context) => {`,
      `  const generatedFunction = true`,
      `  const errors = []`,
      ...code.map(l => `  ${l}`),
      `  return errors`,
      `}`
    ].join('\n')

    try {
      const functionGenerator = new Function('imports', 'schema', functionBody)
      const validateFunction = functionGenerator(imports, {
        schema,
        validateValue: this.validateValue.bind(this)
      })
      this.optimizedValidate = validateFunction
    } catch (e) {
      throw new Error(`Failed to compile optimized function(${e.message}):\n${functionBody}`)
    }
  }

  public abstract toString(options?: ValidatorExportOptions): string

  protected abstract validateValue(value: unknown, context?: string, options?: ValidateOptions): ValidationFailure[]
}
