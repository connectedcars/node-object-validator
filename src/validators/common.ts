import { ValidationErrorContext } from '../errors'

export type Schema = {
  [key: string]: Validator
}

export interface Validator {
  schema?: Schema
  validate(value: unknown, context?: ValidationErrorContext): Error | null
}
