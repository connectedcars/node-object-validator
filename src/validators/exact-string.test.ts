import { RequiredObject, RequiredUnion } from '..'
import { AssertEqual, ValidatorExportOptions } from '../common'
import { NotExactStringFail, RequiredFail } from '../errors'
import {
  isExactString,
  NullableExactString,
  OptionalExactString,
  OptionalNullableExactString,
  RequiredExactString,
  validateExactString
} from './exact-string'

describe('validateExactString (optimize: %s)', () => {
  describe('validateExactString', () => {
    it('requires value to be exact string', () => {
      const value = 'MyString' as unknown
      expect(validateExactString(value, 'MyString')).toStrictEqual([])
    })
  })

  describe('isExactString', () => {
    it('should cast value to string', () => {
      const value = 'MyString' as unknown
      if (isExactString(value, 'MyString')) {
        expect(true as AssertEqual<typeof value, 'MyString'>).toEqual(true)
      } else {
        fail('did not validate but should')
      }
    })
    it('should fail validation', () => {
      const value = 'string' as unknown
      expect(isExactString(value, 'MyString')).toEqual(false)
    })
  })

  describe('RequiredExactString', () => {
    it('should return a function body', () => {
      const validator = new RequiredExactString('MyString', { optimize: false })
      expect(validator.codeGen('value1', 'validator1')).toMatchSnapshot()
    })

    it('toString, constructor', () => {
      const validator = new RequiredExactString('MyString', { optimize: false })
      const code = validator.toString()
      expect(code).toEqual(`new RequiredExactString('MyString', { optimize: false })`)
    })

    it('toString, typescript', () => {
      const validator = new RequiredExactString('MyString', { optimize: false })
      const code = validator.toString({ types: true })
      expect(code).toEqual("'MyString'")
    })
  })
})

describe.each([false, true])('validateExactString (optimize: %s)', optimize => {
  describe('ExactStringValidator', () => {
    it('should generate code for validation and give same result', () => {
      const validator = new RequiredExactString('MyString', { optimize })
      if (optimize) {
        expect(validator['optimizedValidate']).not.toBeNull()
      } else {
        expect(validator['optimizedValidate']).toBeNull()
      }
      const errors = validator.validate('MyString')
      expect(errors).toEqual([])
    })

    it('should export validator code with options', () => {
      const validator = new RequiredExactString('MyString', { optimize })
      const code = validator.toString()
      if (optimize) {
        expect(code).toEqual(`new RequiredExactString('MyString')`)
      } else {
        expect(code).toEqual(`new RequiredExactString('MyString', { optimize: false })`)
      }
    })

    it('accepts values', () => {
      const validator = new RequiredExactString('MyString', { optimize })
      expect(validator.validate('MyString')).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, 'MyString'>).toEqual(true)
    })

    it('rejects invalid values', () => {
      const validator = new RequiredExactString('MyString', { optimize })
      expect(validator.validate('')).toStrictEqual([new NotExactStringFail('Must strictly equal "MyString"', '')])
      expect(validator.validate('mystring')).toStrictEqual([
        new NotExactStringFail('Must strictly equal "MyString"', 'mystring')
      ])
      expect(validator.validate('MyString ')).toStrictEqual([
        new NotExactStringFail('Must strictly equal "MyString"', 'MyString ')
      ])
      expect(validator.validate(' MyString')).toStrictEqual([
        new NotExactStringFail('Must strictly equal "MyString"', ' MyString')
      ])
      expect(validator.validate('bogus')).toStrictEqual([
        new NotExactStringFail('Must strictly equal "MyString"', 'bogus')
      ])
      expect(validator.validate(null)).toStrictEqual([new NotExactStringFail('Must strictly equal "MyString"', null)])
      expect(true as AssertEqual<typeof validator.tsType, 'MyString'>).toEqual(true)
    })

    it('rejects undefined', () => {
      const validator = new RequiredExactString('MyString', { optimize })
      expect(validator.validate(undefined)).toStrictEqual([new RequiredFail('Is required', undefined)])
    })
  })

  describe('OptionalExactString', () => {
    it('accepts empty value', () => {
      const validator = new OptionalExactString('MyString', { optimize })
      expect(validator.validate('MyString')).toStrictEqual([])
      expect(validator.validate(undefined)).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, 'MyString' | undefined>).toEqual(true)
    })

    it('toString, constructor', () => {
      const validator = new OptionalExactString('MyString', { optimize: false })
      const code = validator.toString()
      expect(code).toEqual(`new OptionalExactString('MyString', { required: false, optimize: false })`)
    })

    it('toString, typescript', () => {
      const validator = new OptionalExactString('MyString', { optimize: false })
      const code = validator.toString({ types: true })
      expect(code).toEqual("'MyString' | undefined")
    })
  })

  describe('NullableExactString', () => {
    it('accepts empty value', () => {
      const validator = new NullableExactString('MyString', { optimize })
      expect(validator.validate('MyString')).toStrictEqual([])
      expect(validator.validate(null)).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, 'MyString' | null>).toEqual(true)
    })

    it('toString, constructor', () => {
      const validator = new NullableExactString('MyString', { optimize: false })
      const code = validator.toString()
      expect(code).toEqual(`new NullableExactString('MyString', { nullable: true, optimize: false })`)
    })

    it('toString, typescript', () => {
      const validator = new NullableExactString('MyString', { optimize: false })
      const code = validator.toString({ types: true })
      expect(code).toEqual("'MyString' | null")
    })
  })

  describe('OptionalNullableExactString', () => {
    it('accepts empty value', () => {
      const validator = new OptionalNullableExactString('MyString', { optimize })
      expect(validator.validate('MyString')).toStrictEqual([])
      expect(validator.validate(undefined)).toStrictEqual([])
      expect(validator.validate(null)).toStrictEqual([])
      expect(true as AssertEqual<typeof validator.tsType, 'MyString' | undefined | null>).toEqual(true)
    })

    it('toString, constructor', () => {
      const validator = new OptionalNullableExactString('MyString', { optimize: false })
      const code = validator.toString()
      expect(code).toEqual(
        `new OptionalNullableExactString('MyString', { required: false, nullable: true, optimize: false })`
      )
    })

    it('toString, typescript', () => {
      const validator = new OptionalNullableExactString('MyString', { optimize: false })
      const code = validator.toString({ types: true })
      expect(code).toEqual("'MyString' | undefined | null")
    })
  })
})

describe('Rust Types', () => {
  let typeDefinitions: Record<string, string>
  let options: ValidatorExportOptions

  beforeEach(() => {
    typeDefinitions = {}
    options = {
      types: true,
      language: 'rust',
      typeDefinitions
    }
  })

  it('Required (in a union)', () => {
    const unionValidator = new RequiredUnion([new RequiredExactString(`computerKatten`)], { typeName: 'NeededUnion' })
    expect(unionValidator.toString(options)).toEqual('NeededUnion')

    const expectedNeededUnion = `#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum NeededUnion {
    ComputerKatten,
}

`
    expect(typeDefinitions).toEqual({
      NeededUnion: expectedNeededUnion
    })
  })

  it('Required (in a union), rename', () => {
    const unionValidator = new RequiredUnion([new RequiredExactString(`computerKatten`, {typeName: 'JaMand'})], { typeName: 'NeededUnion' })
    expect(unionValidator.toString(options)).toEqual('NeededUnion')

    const expectedNeededUnion = `#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum NeededUnion {
    #[serde(rename = "JaMand")]
    ComputerKatten,
}

`
    expect(typeDefinitions).toEqual({
      NeededUnion: expectedNeededUnion
    })
  })

  it('Required, Err, in object without a union (taggedunion member, no parent)', () => {
    const objValidator = new RequiredObject({ a: new RequiredExactString(`computerKatten`) }, { typeName: 'ObjName' })

    expect(() => {
      objValidator.toString(options)
    }).toThrow(`ExactString in an object, has to be part of a taggedUnion. str: computerKatten`)
  })

  it('Required, Err, alone', () => {
    const validator = new RequiredExactString(`computerKatten`)

    expect(() => {
      validator.toString(options)
    }).toThrow(`ExactString has to be in an object/union. str: computerKatten`)
  })

  it('Option', () => {
    expect(() => {
      new OptionalExactString('computerKatten').toString(options)
    }).toThrow(`Rust does not support optional ExactString`)

    expect(() => {
      new NullableExactString('computerKatten').toString(options)
    }).toThrow(`Rust does not support optional ExactString`)

    expect(() => {
      new OptionalNullableExactString('computerKatten').toString(options)
    }).toThrow(`Rust does not support optional ExactString`)
  })

  it('Unknown Language', () => {
    expect(() => {
      new RequiredExactString('computerKatten').toString({ types: true, language: 'bingo' as any })
    }).toThrow(`Language: 'bingo' unknown`)
  })
})
