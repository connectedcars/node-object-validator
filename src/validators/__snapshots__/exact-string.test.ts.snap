// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateExactString (optimize: %s) RequiredExactString should return an function body 1`] = `
Array [
  Object {
    "NotExactStringFail": [Function],
    "RequiredFail": [Function],
  },
  Array [],
  Array [
    "const value1 = value1",
    "if (value1 !== undefined) {",
    "  if (value1 !== \\"MyString\\") {",
    "    errors.push(new NotExactStringFail(\`Must strictly equal \\"MyString\\"\`, value1, context))",
    "  }",
    "} else {",
    "  errors.push(new RequiredFail(\`Is required\`, value1, context))",
    "}",
  ],
]
`;
